lisa/tools/ethtool.py

```diff
@@ -0,0 +1,199 @@
+    #       rx-checksumming: on
+    #       tx-checksumming: on
+    #           tx-checksum-ipv4: on
+    #           tx-checksum-ip-generic: off [fixed]
+    #           tx-checksum-ipv6: on
+    #           tx-checksum-fcoe-crc: off [fixed]
+    #           tx-checksum-sctp: off [fixed]
+    #         scatter-gather: on
+    __feature_info_pattern = re.compile(
+        r"Features for (?P<interface>[\w]*):[\s]*(?P<value>.*?)$", re.DOTALL
+    )
+    __feature_settings_pattern = re.compile(
+        r"^[\s]*(?P<name>.*):(?P<value>.*?)?$", re.MULTILINE
+    )
+
+    def __init__(self, device: str, device_feature_raw: str) -> None:
+        self._parse_feature_info(device, device_feature_raw)
+
+    def _parse_feature_info(self, device: str, raw_str: str) -> None:
+        matched_features_info = self.__feature_info_pattern.match(raw_str)
+        if not matched_features_info:
+            raise LisaException("Cannot get device's features settings info")
+
+        self.device_name = device
+        self.enabled_features = []
+        for row in matched_features_info.group("value").splitlines():
+            feature_info = self.__feature_settings_pattern.match(row)
+            if not feature_info:
+                raise LisaException(
+                    "Could not get feature setting in the defined pattern."
+                )
+            if "on" in feature_info.group("value"):
+                self.enabled_features.append(feature_info.group("name"))
+
+
+class Ethtool(Tool):
+    @property
+    def command(self) -> str:
+        return "ethtool"
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        self._command = "ethtool"
+        self._device_list: List[str] = []
+        self._device_channel_map: Dict[str, DeviceChannel] = {}
+        self._device_features_map: Dict[str, DeviceFeatures] = {}
+
+    def _install(self) -> bool:
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages("ethtool")
+        return self._check_exists()
+
+    def _get_device_list(self) -> None:
+        if self._device_list:
+            return
+
+        result = self.node.execute(
+            "test -d /sys/devices && find /sys/devices -name 'net' -a -ipath '*vmbus*'"
+            "-print0"
+        )
+        if result.exit_code != 0:
+            raise LisaException(
+                "Cannot find synthetic network interface. No /sys/devices directory. "
+                f"exit_code: {result.exit_code} "
+                f"stderr: {result.stderr}"
+            )
+        # more than one path can be found for "net" folder.
+        # split the command output on null terminator.
+        netdirs = result.stdout.split("\x00")
+        for netdir in netdirs:
+            if not netdir:
+                continue
+            cmd_result = self.node.execute(f"ls {PurePath(netdir)} | head -n 1")
+            if cmd_result.exit_code != 0:
```

Comment:
The logic is to check if there is any file under a folder. I think we need another tool ls too. But it's ok to use it here this time.
The cmd can be simpler like `ls {netdir}`, and check if there is output. It doesn't need to head. I tested on an Ubuntu, the exit code is zero, if nothing returned in a folder. So it looks not effective.

Reply:
It needs the head because I want to grab only the interface name from the "ls" output. You must have noticed ls {netdir} would give details about permissions etc etc. I am tryign to grab only the file name